### PR TITLE
feat: report live run phases for dashboard visibility

### LINKED ISSUE #16572: [Feature]: Show real-time automation run phases in the dashboard
### Problem or Motivation

Users need to understand what an automation is doing while it is running, not only after it finishes.

A generic running state is not enough for longer or multi-step automations. For example, a code-review automation may move through phases like polling recent tagged PRs, checking out a PR, examining the diff, QAing changes, and posting review comments. Without phase-level progress, users cannot tell whether the automation is making progress, stuck, blocked, or close to completion.

Existing visibility work covers health state, disabled/blocking reasons, conversation links, and broader monitoring. This ticket focuses on live, user-facing phase/status updates during an automation run.

### Desired Behavior

As a user viewing an automation run or automation dashboard, I can see concise real-time status phases that describe what the run is currently doing.

Automation runs should expose a phase/status message that can update while the run is active. The dashboard should receive and render those updates without requiring a full page refresh. Preset automations should emit standard phase updates where possible, and custom automations should have a documented way to emit safe user-facing phase messages.

Example phases for a code-review automation:

1. Polling recent PRs with the target label
2. Checking out PR #123
3. Examining the diff
4. Running QA checks
5. Posting review comments

### Acceptance Criteria

- [ ] Automation runs expose a phase/status message field that can update while the run is active.
- [ ] Automation dashboard/run detail UI renders live phase updates without requiring a full page refresh.
- [ ] Preset automations can emit standard phase updates for common run steps.
- [ ] Custom automations have a documented way to emit phase updates.
- [ ] Phase messages are concise, user-facing, and safe to show in dashboard surfaces.
- [ ] Missing phase information degrades gracefully to existing run status.
- [ ] Phase updates do not expose secrets, raw payloads, stack traces, or sensitive repository/customer data.
- [ ] A screenshot or video demonstrating live phase updates during an automation run is attached to the PR.

### Alternatives Considered

- Show only `running`, `completed`, or `failed`. This is simpler, but it does not explain what is happening during a long-running automation.
- Rely on conversation logs for progress. Logs have detail, but they require users to leave the dashboard and inspect a lower-level view.
- Show phase updates only after completion. This helps summaries, but it does not improve confidence while a run is active.

### Additional Context

This comes from the Automation Funnel Optimization work. After creation, the product needs to improve execution visibility and recovery. Real-time phase updates can make automations feel observable and trustworthy rather than opaque background jobs.

Related work:

- [PRD-146](https://linear.app/all-hands-ai/issue/PRD-146/customer-feature-request-automation-monitoring-and-debugging) — Customer Feature Request: Automation monitoring and debugging
- [OpenHands/OpenHands#16494](https://github.com/OpenHands/OpenHands/issues/16494) — Surface unhealthy/disabled automations and their blocking reason in the dashboard
- [OSS-9492](https://linear.app/all-hands-ai/issue/OSS-9492/register-automation-run-conversation-id-as-soon-as-conversation-is) — Register automation run conversation ID as soon as conversation is created
- [OpenHands/OpenHands#16569](https://github.com/OpenHands/OpenHands/issues/16569) — Add automation run-summary metadata and surface latest summaries in the dashboard

---

_This issue was created by an AI agent (OpenHands) on behalf of the user._

### PR DESCRIPTION
**Pull request by hieptl**
## Why

Users cannot tell what an automation run is doing while it is active — the API exposes only a coarse run status until the run terminates, so long or multi-step automations (e.g. a code-review automation that clones a repo, examines a diff, and posts comments) feel opaque: there is no way to see whether a run is progressing, stuck, or nearly done. Issue OpenHands/OpenHands#16572 asks for live, user-facing phase updates during a run.

There was no mid-run write channel at all: the only sandbox→service writes were the terminal `/complete` and `/cancel` endpoints, and the preset scripts printed their phase banners to stdout only.

## Summary

- Add `AutomationRun.current_phase` (nullable, 200 chars; migration `021`, revising `020`) plus a sandbox-callable `POST /v1/runs/{run_id}/phase` endpoint that reuses the completion callback's auth/ownership pattern, only accepts writes while the run is PENDING/RUNNING (409 once terminal), normalizes the message server-side, and exposes the field on `AutomationRunResponse`.
- Instrument the run pipeline end to end: the dispatcher injects `AUTOMATION_PHASE_URL` and writes "Preparing environment" / "Starting automation" infra phases; the preset `sdk_main.py` scripts report standard setup phases and then stream redacted, throttled LLM-authored per-action summaries as live phases; both `setup.sh` scripts report "Installing dependencies". All reporting is best-effort — a failed phase POST can never fail a run.
- Document the contract for custom automations in `docs/run-phase-reporting.md` (env vars, curl/Python examples, response codes, safety guidance: no secrets, payloads, or stack traces in phase text).

## Issue Number

Fixes OpenHands/OpenHands#16572 (backend half; the dashboard rendering lands in the GUI repository).

## How to Test

1. Run the suite: `make test` (or `uv run python -m pytest tests/ -q`) — see the new `TestReportRunPhase` and dispatcher phase tests.
2. Manual API check against a local service:
   - Dispatch a run, then while it is RUNNING:
     `curl -X POST "$BASE/v1/runs/<run_id>/phase" -H "Authorization: Bearer <key>" -H "Content-Type: application/json" -d '{"phase": "Examining the diff"}'`
   - `GET /v1/{automation_id}/runs` now returns `current_phase` on the run.
   - Repeat the POST after the run completes → `409`.
3. Create a **new** prompt-preset automation (tarballs are baked at creation; pre-existing automations keep their old scripts) and dispatch it: the run's `current_phase` advances through Preparing environment → Installing dependencies → Setting up workspace → … → Agent is working on the task → live action summaries.

## Video/Screenshots

Live-dashboard screencast is attached to the companion GUI PR (per the issue's acceptance criteria); this PR is API/pipeline only. Reviewer-reproducible evidence here: the `TestReportRunPhase` suite and the curl flow above.

## Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Docs / chore

### CODE REVIEW

File: openhands/automation/router.py [hunk 1]
```diff
@@ -604,6 +605,59 @@ async def complete_run(
...
+        .where(
+            AutomationRun.id == run_id,
+            AutomationRun.status.in_(
+                (AutomationRunStatus.PENDING, AutomationRunStatus.RUNNING)
+            ),
+        )
+        .values(current_phase=body.phase)
+    )
+    db_result: CursorResult = await session.execute(stmt)  # type: ignore[assignment]
+    if db_result.rowcount == 0:
```
Comment (tofarr) [inline openhands/automation/router.py:732]: This 409 path mirrors `complete_run`'s optimistic-locking pattern nicely. One small gap: `TestReportRunPhase` covers the 409 (terminal run) case but there's no test for the 404 (`run is None`) branch. A one-liner test posting to a random UUID would close that coverage gap.

File: openhands/automation/presets/prompt/sdk_main.py [hunk 2]
```diff
@@ -128,6 +129,60 @@
...
+        )
+
+        return redact_api_key_literals(redact_text_secrets(text))
+    except Exception:
+        return ""
+
+
+# Latest pending live phase; a daemon thread posts it at most once per
+# interval so the conversation event thread never blocks on network I/O.
+_live_phase: dict = {"pending": None, "posted": None, "stop": False}
```
Comment (tofarr) [inline openhands/automation/presets/prompt/sdk_main.py:177]: This cross-thread dict relies on CPython GIL atomicity for the `pending`/`posted`/`stop` reads/writes. That's fine for this last-writer-wins cosmetic telemetry pattern (worst case: a slightly stale phase posts on the next interval), but a one-line comment noting the intentional GIL reliance would help future readers who might reach for a `threading.Lock`.

File: openhands/automation/presets/prompt/sdk_main.py [hunk 3]
```diff
@@ -128,6 +129,60 @@
...
+AUTOMATION_PHASE_URL = os.environ.get("AUTOMATION_PHASE_URL", "")
+_PHASE_TOKEN = (
+    os.environ.get("AUTOMATION_CALLBACK_API_KEY")
+    or os.environ.get("OPENHANDS_API_KEY")
+    or ""
+)
+PHASE_POST_INTERVAL_SECONDS = 4.0
+
+
+def report_phase(message: str) -> None:
```
Comment (tofarr) [inline openhands/automation/presets/prompt/sdk_main.py:143]: This `report_phase` / `_redact_phase` / `_live_phase` / `_phase_poster` block is duplicated verbatim in `presets/plugin/sdk_main.py`. I understand these are self-contained preset templates baked into tarballs at creation time (and the file already carries a "keep in sync with presets/prompt/sdk_main.py" note for finish-tool wiring). Flagging it as a maintainability note — if this logic evolves, consider shipping a shared helper module alongside the entrypoint so both presets import it rather than copy it.

File: openhands/automation/schemas.py [hunk 4]
```diff
@@ -773,6 +776,22 @@ def parse_sdk_conversation_error(cls, value: Any) -> Any:
             return value
 
 
+class RunPhaseRequest(BaseModel):
```
Comment (tofarr) [inline openhands/automation/schemas.py:784]: The `mode="before"` normalization here is the right call — collapsing control chars/whitespace server-side means clients can't inject newlines into the dashboard regardless of how they build the payload. `extra="forbid"` matches the rest of the schemas. Clean.

**Review by tofarr — COMMENTED** [review]
## Review Summary

This is a well-designed, well-tested feature PR. The core API, migration, and dispatcher wiring are solid — the best-effort philosophy is consistently applied across the service, dispatcher, preset scripts, and setup.sh. CI passes (backend + unit-tests). I'm leaving a **COMMENT** review with a few minor suggestions; nothing here blocks merge.

### Strengths

- **Auth/ownership reuse** — `report_run_phase` faithfully mirrors `complete_run`'s proven pattern (`_require_manage_automations`, ownership check, 403 on mismatch).
- **Optimistic locking** on `status IN (PENDING, RUNNING)` with 409 on `rowcount == 0` prevents races between phase writes and terminal transitions.
- **Input normalization** server-side (`RunPhaseRequest.normalize_phase`) collapses control chars/whitespace runs — clients can't inject multiline content into the dashboard.
- **Security** — `_redact_phase` drops the phase entirely (returns `""`) if redaction fails or strips anything, and the docs explicitly warn against secrets/payloads/stack traces.
- **`current_phase` deliberately not cleared on completion** (unlike `status_detail`) — a documented design decision that lets the dashboard show the last phase briefly after the run ends.
- **Migration** is correct: revision 020 revises 019, nullable column, SQLite-safe (skips `COMMENT`), working downgrade.

### Minor observations (non-blocking)

1. **PR description says "migration `017`"** but the actual file is `020_add_run_current_phase.py` (revises 019). Just a stale reference in the body; the code is correct.
2. **No test for the 404 path** (`run is None`). The 404 branch in `report_run_phase` is untested in `TestReportRunPhase`. It's simple and mirrors `complete_run`, but a one-liner test would close the coverage gap.
3. **Duplicated phase-reporting code** between `presets/plugin/sdk_main.py` and `presets/prompt/sdk_main.py` (~71 lines verbatim each). I understand these are self-contained preset templates baked into tarballs at creation time and the duplication is an existing pattern (the file even has a "keep in sync" comment for finish-tool wiring). Still, a shared helper shipped alongside the entrypoint would reduce the maintenance surface if this logic evolves.
4. **`_live_phase` dict relies on GIL atomicity** for cross-thread reads/writes of `pending`/`posted`/`stop`. This is safe in CPython for this "last-writer-wins, single reader" cosmetic pattern, but a one-line comment noting the intentional reliance on GIL would help future readers.
5. **`update_run_current_phase` failure path** (the `except: logger.exception` branch) is untested. Low priority since it's a trivial best-effort wrapper, but worth noting.

### Risk assessment
- **Security**: LOW — proven auth pattern, ownership enforced, input normalized, LLM summaries redacted.
- **Correctness**: LOW — optimistic locking, best-effort design, thorough tests.
- **Maintainability**: MEDIUM-LOW (preset-script duplication is an accepted pattern here).

_This review was created by an AI agent (OpenHands) on behalf of the reviewer._

**Review by tofarr — APPROVED** [review]
🍰

File: openhands/automation/router.py (hunk 1 above)
Comment (hieptl) [reply openhands/automation/router.py:732]: Good catch — added `test_report_phase_for_unknown_run_returns_404` in fc0f347 (posts to a random UUID and asserts 404).

File: openhands/automation/presets/prompt/sdk_main.py (hunk 2 above)
Comment (hieptl) [reply openhands/automation/presets/prompt/sdk_main.py:177]: Added a note on `_live_phase` in both preset scripts (fc0f347): the dict is deliberately lock-free because each key is a single-reference read/write (atomic under the GIL) and last-writer-wins is acceptable for cosmetic telemetry — so nobody reaches for a `threading.Lock` later without knowing it was considered.

File: openhands/automation/presets/prompt/sdk_main.py (hunk 3 above)
Comment (hieptl) [reply openhands/automation/presets/prompt/sdk_main.py:143]: Agreed on the maintenance surface. I kept the block inline for this PR because preset tarballs are assembled from exactly two files — `_load_*_preset_files` reads `sdk_main.py` → `main.py` plus `setup.sh` — so a shared helper would mean extending the packaging contract (the cache loaders, the tarball writer, and the layout the sandbox unpacks) for both presets. That felt like its own change rather than something to fold in here.

To make the duplication explicit in the meantime, both blocks now carry a "keep this block in sync with presets/<other>/sdk_main.py" note (fc0f347), matching the existing finish-tool convention. If a third preset or more phase logic shows up, I am happy to do the shared-module refactor as a follow-up.

**Comment by hieptl** [comment]
Thanks for the review! Addressed in fc0f347:

- **Migration collision** — after the merge from `main`, `020` was already taken by `020_add_automation_disabled_reason` (#352), which is what broke `unit-tests` (`Revision 020 is present more than once`). The `current_phase` migration is now `021` (revises `020`); `tests/test_db.py` passes locally again.
- **404 path** — covered by `test_report_phase_for_unknown_run_returns_404`.
- **`update_run_current_phase` failure path** — covered by `TestUpdateRunCurrentPhase::test_database_failure_is_logged_not_raised` (session factory raises → logged, never raised).
- **GIL note** — added on `_live_phase` in both preset scripts.
- **Duplication** — kept inline for now with an explicit keep-in-sync note; reasoning in the thread.
- **PR description** — now references migration `021` instead of `017`.
